### PR TITLE
facts.pp: use localhost instead of 127.0.0.1

### LINKED ISSUE #330: Jira_facts deprecation warnings
# Affected Puppet, Ruby, OS and module versions/distributions

- Puppet: 6.18.0
- Ruby: ruby 2.7.0p0 (2019-12-25 revision 647ee6f091) 
- Distribution: Ubuntu 20.04
- Module version: 4.0.1

## How to reproduce (e.g Puppet code you use)

```
class { 'jira::facts': }
```
## What are you seeing

```
Warning: Facter: external fact file "/etc/facter/facts.d/jira_facts.rb" had output on stderr: /etc/facter/facts.d/jira_facts.rb:17: warning: calling URI.open via Kernel#open is deprecated, call URI.open directly or use URI#open
```
## What behaviour did you expect instead

No warnings

## Output log

```
Warning: Facter: external fact file "/etc/facter/facts.d/jira_facts.rb" had output on stderr: /etc/facter/facts.d/jira_facts.rb:17: warning: calling URI.open via Kernel#open is deprecated, call URI.open directly or use URI#open
Info: Using configured environment 'CustomerInfra'
Info: Retrieving pluginfacts
Info: Retrieving plugin
Info: Retrieving locales
Info: Loading facts
Warning: Facter: external fact file "/etc/facter/facts.d/jira_facts.rb" had output on stderr: /etc/facter/facts.d/jira_facts.rb:17: warning: calling URI.open via Kernel#open is deprecated, call URI.open directly or use URI#open
Info: Caching catalog for jira01-qual.node.customer.cloud
Notice: Applied catalog in 14.88 seconds
```

### PR DESCRIPTION
**Pull request by kenyon**
Use localhost instead of 127.0.0.1 so that we don't discriminate against IPv6-only configurations.

### CODE REVIEW

**Review by ekohl — APPROVED** [review]

**Comment by kenyon** [comment]
Needs #340 for the tests to pass.

**Comment by kenyon** [comment]
I split this PR into separate PRs so that we produce a better changelog.

File: spec/classes/jira_facts_spec.rb [hunk 1]
```diff
@@ -10,8 +10,7 @@
 
         regexp_pe = %r{^#\!/opt/puppet/bin/ruby$}
         regexp_oss = %r{^#\!/opt/puppetlabs/puppet/bin/ruby$}
-        regexp_url = %r{http://127.0.0.1\:8080/rest/api/2/serverInfo}
-        pe_external_fact_file = '/etc/puppetlabs/facter/facts.d/jira_facts.rb'
+        regexp_url = Regexp.new(Regexp.escape('http://localhost:8080/rest/api/2/serverInfo'))
```
Comment (ekohl) [inline spec/classes/jira_facts_spec.rb:13]: Just a thought: should it be
```suggestion
        url = 'http://localhost:8080/rest/api/2/serverInfo'
```

And then later `with_content(Regexp.new(Regexp.escape(url)))`

File: spec/classes/jira_facts_spec.rb [hunk 2]
```diff
@@ -24,7 +23,7 @@
           end
 
           it do
-            is_expected.to contain_file(pe_external_fact_file). \
+            is_expected.to contain_file(external_fact_file). \
```
Comment (ekohl) [inline spec/classes/jira_facts_spec.rb:26]: Perhaps this should be kept to a separate PR and remove the specifics for Puppet Enterprise? I think `$facts['aio_agent_version']` should also cover Puppet Enterprise. Then the specific testcase can also be removed.

File: manifests/facts.pp [hunk 3]
```diff
@@ -21,12 +21,12 @@
   $json_packages = $jira::params::json_packages,
   # lint:ignore:parameter_order
   $uri           = $jira::tomcat_address ? {
-    undef   => '127.0.0.1',
+    undef   => 'localhost',
     default => $jira::tomcat_address,
   },
```
Comment (ekohl) [inline manifests/facts.pp:26]: You can clean this up to:
```suggestion
  $uri = pick($jira::tomcat_address, 'localhost')
```

**Review by ekohl — COMMENTED** [review]
We should also squash commits to keep the history a bit cleaner.

File: spec/classes/jira_facts_spec.rb (hunk 2 above)
Comment (kenyon) [reply spec/classes/jira_facts_spec.rb:26]: I see what you mean, that sounds good.

**Review by ekohl — COMMENTED** [review]
Could you rebase this now that modulesync was merged?

**Comment by ekohl** [comment]
I opened https://github.com/voxpupuli/puppet-jira/pull/347. Can we merge that first and then rebase this on top of it?

**Comment by ekohl** [comment]
I opened https://github.com/voxpupuli/puppet-jira/pull/349 to fix the tests

**Pull request merged**